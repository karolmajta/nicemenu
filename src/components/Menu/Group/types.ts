import { IconName } from "../Item/Icon/types";
import React from "react";
import { AllowedShortcutKey, ItemProps } from "../Item/types";

export interface GroupItemProps {
  icon: IconName;
  label: string;
  shortcut?: Array<Array<AllowedShortcutKey>>;
  goto?: string;
  onSelect?: () => void;
  listIndex: number;
}

export interface GroupProps {
  label: string;
  items: Array<GroupItemProps>;
  activeListIndex: number | null;
  className?: string;
  classNames?: Array<string>;
  components?: {
    Item?: React.ComponentType<ItemProps>;
  };
  onItemMouseMove?: (index: number) => void;
  onGoto: (gotoPath: string) => void;
}
