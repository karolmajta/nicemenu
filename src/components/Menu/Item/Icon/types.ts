export type IconName =
  | "bolt"
  | "dot"
  | "inbox"
  | "loader"
  | "pencil"
  | "reschedule";

export interface IconProps {
  name: IconName;
  className?: string;
  classNames?: Array<string>;
}
