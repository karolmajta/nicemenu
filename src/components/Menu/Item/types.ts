import React from "react";
import { IconName, IconProps } from "./Icon/types";

export type AllowedShortcutKey =
  | "a"
  | "b"
  | "c"
  | "d"
  | "e"
  | "f"
  | "g"
  | "h"
  | "i"
  | "j"
  | "k"
  | "l"
  | "m"
  | "n"
  | "o"
  | "p"
  | "q"
  | "r"
  | "s"
  | "t"
  | "u"
  | "v"
  | "w"
  | "x"
  | "y"
  | "z"
  | "Shift";

export interface ItemProps {
  label: string;
  icon?: IconName;
  shortcut?: Array<Array<AllowedShortcutKey>>;
  listIndex: number;
  highlighted: boolean;
  goto?: string;
  className?: string;
  classNames?: Array<string>;
  components?: {
    Icon?: React.ComponentType<IconProps>;
  };
  onMouseMove?: (listIndex: number) => void;
  onSelect?: () => void;
  onGoto: (menuKey: string) => void;
}
