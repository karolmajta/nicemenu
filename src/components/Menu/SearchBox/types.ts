export interface SearchBoxProps {
  placeholder: string;
  searchTerm: string;
  focused: boolean;
  className?: string;
  classNames?: Array<string>;
  onArrowUp: () => void;
  onArrowDown: () => void;
  onChange: (searchTerm: string) => void;
  onFocus: () => void;
}
