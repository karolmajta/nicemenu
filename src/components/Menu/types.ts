import { IconName } from "./Item/Icon/types";
import React from "react";
import { ItemProps, AllowedShortcutKey } from "./Item/types";
import { GroupProps } from "./Group/types";
import { SearchBoxProps } from "./SearchBox/types";

export interface MenuItem {
  icon: IconName;
  label: string;
  shortcut?: Array<Array<AllowedShortcutKey>>;
  goto?: string;
  onSelect?: () => void;
}

export interface MenuGroup {
  label: string;
  items: Array<MenuItem>;
}

export interface MenuProps {
  title: string;
  items: Array<MenuItem>;
  groups: Array<MenuGroup>;
  submenus: {
    [submenuName: string]: MenuProps;
  };
  className?: string;
  classNames?: Array<string>;
  components?: {
    Item?: React.ComponentType<ItemProps>;
    Group?: React.ComponentType<GroupProps>;
    SearchBox?: React.ComponentType<SearchBoxProps>;
  };
}
