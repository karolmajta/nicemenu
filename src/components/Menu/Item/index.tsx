import React, { useEffect, useState, useContext } from "react";
import cx from "classnames";
import styled from "styled-components";

import WindowContext from "../../../contexts/WindowContext";
import { ItemProps } from "./types";
import { IGNORED_KEYS, KEY_SEPARATOR } from "./utils";
import Icon, { IconUnstyled } from "./Icon";

export function ItemUnstyled({
  label,
  icon,
  shortcut,
  className,
  classNames,
  components,
  listIndex,
  highlighted,
  goto,
  onMouseMove,
  onSelect,
  onGoto,
}: ItemProps) {
  const [currentlyPressedKeys, setCurrentlyPressedKeys] = useState<Set<string>>(
    new Set()
  );
  const window = useContext(WindowContext);

  const [shortcutIndex, setShortcutIndex] = useState(0);

  const IconComponent = components?.Icon || IconUnstyled;

  useEffect(() => {
    const onKeyUp = (e: KeyboardEvent) => {
      if (highlighted && e.key === "Enter" && onSelect) {
        onSelect();
      }
      if (highlighted && e.key === "Enter" && goto) {
        goto && onGoto(goto);
      }
    };
    window.addEventListener("keyup", onKeyUp);

    return () => {
      window.removeEventListener("keyup", onKeyUp);
    };
  }, [highlighted, onSelect, goto, onGoto, window]);

  useEffect(() => {
    const onKeyDown = (e: KeyboardEvent) => {
      if (
        !shortcut ||
        currentlyPressedKeys.has(e.key) ||
        IGNORED_KEYS.has(e.key)
      ) {
        setCurrentlyPressedKeys(new Set());
        return;
      }
      const newKeys = new Set([
        ...Array.from(currentlyPressedKeys),
        e.key.toLowerCase(),
      ]);
      const expectedKeys = shortcut[shortcutIndex];
      setCurrentlyPressedKeys(newKeys);
      if (
        newKeys.size > currentlyPressedKeys.size &&
        expectedKeys.every((key) => newKeys.has(key.toLowerCase()))
      ) {
        const expectedTogetherNow = Array.from(
          shortcut[shortcutIndex].map((k) => k.toLowerCase())
        )
          .sort()
          .join(KEY_SEPARATOR);
        const presentTogetherNow = Array.from(newKeys)
          .sort()
          .join(KEY_SEPARATOR);
        if (shortcutIndex + 1 === shortcut.length) {
          setShortcutIndex(0);
          if (expectedTogetherNow === presentTogetherNow) {
            onSelect && onSelect();
            goto && onGoto(goto);
          }
        } else {
          if (expectedTogetherNow === presentTogetherNow) {
            setShortcutIndex((i) => i + 1);
          } else {
            setShortcutIndex(0);
          }
        }
      }
    };

    const onKeyUp = (e: KeyboardEvent) => {
      if (!shortcut || IGNORED_KEYS.has(e.key)) {
        setCurrentlyPressedKeys(new Set());
        return;
      }
      setCurrentlyPressedKeys(
        (keys) =>
          new Set(Array.from(keys).filter((key) => key !== e.key.toLowerCase()))
      );
      let nonResetKeysArray: Array<string> = [];
      shortcut.slice(0, shortcutIndex).forEach((keys) => {
        nonResetKeysArray = [
          ...nonResetKeysArray,
          ...keys.map((k) => k.toLowerCase()),
        ];
      });
      const nonResetKeys = new Set(nonResetKeysArray);
      if (!nonResetKeys.has(e.key.toLowerCase())) {
        setShortcutIndex(0);
      }
    };

    window.addEventListener("keydown", onKeyDown);
    window.addEventListener("keyup", onKeyUp);

    return () => {
      window.removeEventListener("keydown", onKeyDown);
      window.removeEventListener("keyup", onKeyUp);
    };
  }, [
    shortcut,
    shortcutIndex,
    setShortcutIndex,
    currentlyPressedKeys,
    setCurrentlyPressedKeys,
    onSelect,
    goto,
    onGoto,
    window,
  ]);

  return (
    <div
      data-testid="Laserfocus-Menu-Item"
      onMouseMove={() => onMouseMove && onMouseMove(listIndex)}
      onClick={(e) => {
        onSelect && onSelect();
        goto && onGoto(goto);
        e.stopPropagation();
      }}
      className={cx(
        "Laserfocus-Menu-Item",
        { highlighted },
        className,
        ...(classNames || [])
      )}
    >
      {icon ? (
        <IconComponent name={icon} />
      ) : (
        <div className="icon-placeholder" />
      )}
      <div className="item-label">{label}</div>
      {shortcut && (
        <div className="item-shortcut">
          {shortcut.map((steps, i, shortcut) => (
            <div key={i} className="item-shortcut-parts">
              {steps.map((step, j) => (
                <div key={j} className="item-shortcut-step">
                  {step}
                </div>
              ))}
              {i < shortcut.length - 1 && (
                <div className="item-shortcut-part-separator">then</div>
              )}
            </div>
          ))}
        </div>
      )}
    </div>
  );
}

const Item = styled(ItemUnstyled).attrs((props) => ({
  ...props,
  components: {
    Icon: Icon,
  },
}))`
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;

  width: 100%;
  height: 44px;

  cursor: pointer;

  &.highlighted {
    background: rgba(228, 228, 238, 0.05);
  }

  .icon-placeholder {
    margin-left: 46px;
  }

  .item-label {
    display: flex;
    flex-direction: row;
    justify-content: flex-start;
    align-items: center;

    width: 100%;
    height: 36px;

    font-size: 15px;
    color: rgba(228, 228, 238, 0.8);
  }

  .item-shortcut {
    display: flex;
    flex-direction: row;
    justify-self: flex-end;

    margin-right: 20px;

    .item-shortcut-parts {
      display: flex;
      flex-direction: row;

      .item-shortcut-step {
        display: block;
        padding: 0 5px;

        height: 20px;
        min-width: 10px;
        margin-left: 12px;

        text-align: center;
        font-weight: 500;
        font-size: 12px;
        line-height: 20px;
        color: #828282;

        background: rgba(228, 228, 238, 0.1);
        border-radius: 4px;
      }

      .item-shortcut-part-separator {
        display: block;

        height: 20px;
        margin-left: 12px;

        font-weight: 500;
        font-size: 12px;
        line-height: 20px;
        color: #828282;
      }
    }
  }
`;

export default Item;
