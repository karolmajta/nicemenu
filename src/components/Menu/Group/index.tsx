import React from "react";
import cx from "classnames";
import styled from "styled-components";
import Item, { ItemUnstyled } from "../Item";
import { GroupProps } from "./types";

export function GroupUnstyled({
  label,
  items,
  activeListIndex,
  className,
  classNames,
  components,
  onItemMouseMove,
  onGoto,
}: GroupProps) {
  const ItemComponent = components?.Item || ItemUnstyled;
  return (
    <div
      data-testid="Laserfocus-Menu-Group"
      className={cx("Laserfocus-Menu-Group", className, ...(classNames || []))}
    >
      <div className="group-label">
        <div className="group-label-text">{label}</div>
      </div>
      {items.map(({ icon, label, shortcut, listIndex, goto, onSelect }, i) => (
        <ItemComponent
          key={i}
          icon={icon}
          label={label}
          shortcut={shortcut}
          listIndex={listIndex}
          highlighted={listIndex === activeListIndex}
          onMouseMove={() => onItemMouseMove && onItemMouseMove(listIndex)}
          goto={goto}
          onGoto={onGoto}
          onSelect={() => onSelect && onSelect()}
        />
      ))}
    </div>
  );
}

const Group = styled(GroupUnstyled).attrs((props) => ({
  ...props,
  components: {
    Item: Item,
  },
}))`
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;

  width: 100%;

  .group-label {
    display: flex;
    flex-direction: row;
    justify-content: flex-start;
    align-items: center;

    width: 100%;
    height: 36px;

    font-size: 12px;
    color: rgba(228, 228, 238, 0.5);

    background: rgba(228, 228, 238, 0.05);

    .group-label-text {
      display: block;

      margin-left: 20px;
    }
  }
`;

export default Group;
