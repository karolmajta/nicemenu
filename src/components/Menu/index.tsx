import React, { useState, useEffect, useContext } from "react";
import styled from "styled-components";
import cx from "classnames";

import Item, { ItemUnstyled } from "./Item";
import Group, { GroupUnstyled } from "./Group";
import SearchBox, { SearchBoxUnstyled } from "./SearchBox";
import WindowContext from "../../contexts/WindowContext";
import { MenuProps } from "./types";
import { addListIndexes, performSearch, buildNewStack } from "./utils";

export function MenuUnstyled({
  title,
  items,
  groups,
  submenus,
  className,
  classNames,
  components,
}: MenuProps) {
  const window = useContext(WindowContext);

  const [activeListIndex, setActiveListIndex] = useState<number | null>(null);
  const [submenuStack, setSubmenuStack] = useState<Array<string>>([]);
  const [searchTerm, setSearchTerm] = useState("");

  const ItemComponent = components?.Item || ItemUnstyled;
  const GroupComponent = components?.Group || GroupUnstyled;
  const SearchBoxComponent = components?.SearchBox || SearchBoxUnstyled;

  let activeSubmenus = submenus;
  let activeItems = items;
  let activeGroups = groups;
  let activeTitle = title;

  submenuStack.forEach((submenuKey) => {
    activeItems = activeSubmenus[submenuKey].items;
    activeGroups = activeSubmenus[submenuKey].groups;
    activeTitle = activeSubmenus[submenuKey].title;
    activeSubmenus = activeSubmenus[submenuKey].submenus;
  });

  if (searchTerm.trim()) {
    activeItems = performSearch(searchTerm, items, groups, submenus);
    activeGroups = [];
  }

  const { numberedItems, numberedGroups, listItemsCount } = addListIndexes(
    activeItems,
    activeGroups
  );

  useEffect(() => {
    const onKeyUp = (event: KeyboardEvent) => {
      if (event.key === "ArrowDown") {
        if (activeListIndex === null) {
          setActiveListIndex(0);
        } else if (activeListIndex >= listItemsCount - 1) {
          setActiveListIndex(0);
        } else {
          setActiveListIndex((i) => (i || 0) + 1);
        }
      }
      if (event.key === "ArrowUp") {
        if (activeListIndex === null) {
          setActiveListIndex(listItemsCount - 1);
        } else if (activeListIndex <= 0) {
          setActiveListIndex(listItemsCount - 1);
        } else {
          setActiveListIndex((i) => (i || 0) - 1);
        }
      }
      if (event.key === "ArrowLeft") {
        setActiveListIndex(null);
      }
    };

    window.addEventListener("keyup", onKeyUp);

    return () => {
      window.removeEventListener("keyup", onKeyUp);
    };
  }, [activeListIndex, listItemsCount, window]);

  return (
    <div
      data-testid="Laserfocus-Menu"
      className={cx("Laserfocus-Menu", className, ...(classNames || []))}
    >
      <div className="menu-title">
        <div className="menu-title-text">
          {!searchTerm.trim()
            ? activeTitle
            : `Search results for: ${searchTerm}`}
        </div>
      </div>
      <SearchBoxComponent
        searchTerm={searchTerm}
        placeholder={
          activeListIndex === null
            ? "Search..."
            : "Press arrow left to search or type command..."
        }
        onChange={setSearchTerm}
        focused={activeListIndex === null}
        onArrowDown={() => setActiveListIndex(-1)}
        onArrowUp={() => setActiveListIndex(listItemsCount)}
        onFocus={() => setActiveListIndex(null)}
      />
      {numberedItems.map(
        ({ icon, label, shortcut, listIndex, onSelect, goto }, i) => (
          <ItemComponent
            key={i}
            icon={icon}
            label={label}
            shortcut={shortcut}
            listIndex={listIndex}
            highlighted={activeListIndex === listIndex}
            onMouseMove={(listIndex) => setActiveListIndex(listIndex)}
            onSelect={() => onSelect && onSelect()}
            goto={goto}
            onGoto={(path) => {
              setSubmenuStack((oldStack) => buildNewStack(oldStack, path));
              setActiveListIndex(0);
            }}
          />
        )
      )}
      {numberedGroups.map(({ label, items }, i) => (
        <GroupComponent
          key={i}
          label={label}
          items={items}
          activeListIndex={activeListIndex}
          onItemMouseMove={(listIndex) => setActiveListIndex(listIndex)}
          onGoto={(path) => {
            setSubmenuStack((oldStack) => buildNewStack(oldStack, path));
            setActiveListIndex(0);
          }}
        />
      ))}
    </div>
  );
}

const Menu = styled(MenuUnstyled).attrs((props) => ({
  ...props,
  components: {
    Item: Item,
    Group: Group,
    SearchBox: SearchBox,
  },
}))`
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
  padding: 12px 0;

  font-family: Inter;
  font-style: normal;
  font-weight: normal;

  background: #2d3346;
  box-shadow: 0px 50px 100px -20px rgba(11, 31, 82, 0.27),
    0px 22.3363px 17.869px rgba(72, 136, 159, 0.0417275),
    0px 6.6501px 5.32008px rgba(72, 136, 159, 0.0282725);
  border-radius: 12px;

  .menu-title {
    display: flex;
    flex-direction: row;
    justify-content: flex-start;
    align-items: center;
    align-self: flex-start;

    height: 22px;
    margin-top: -3px;
    margin-left: 20px;

    font-size: 10px;
    font-weight: 500;
    letter-spacing: 1.5px;

    color: #e4e4ee;
    background: rgba(228, 228, 238, 0.05);
    border-radius: 3px;

    .menu-title-text {
      display: block;

      margin: 0 16px;
    }
  }
`;

export default Menu;
