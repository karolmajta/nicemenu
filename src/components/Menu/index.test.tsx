import React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";
import Menu from ".";
import { MenuProps } from "./types";

const defaultTitle = "Main menu";

const defaultItems = [
  {
    icon: "bolt",
    label: "Log an activity",
    shortcut: [["c"]],
    onSelect: () => jest.fn(),
  },
  {
    icon: "reschedule",
    label: "Set a reminder",
    shortcut: [["h"]],
    onSelect: () => jest.fn(),
  },
  {
    icon: "loader",
    label: "Change status",
    shortcut: [["Shift", "s"]],
    onSelect: () => jest.fn(),
  },
  {
    icon: "pencil",
    label: "Edit contact",
    shortcut: [["e"]],
    onSelect: () => jest.fn(),
  },
];

const defaultGroups = [
  {
    label: "Go to...",
    items: [
      {
        icon: "inbox",
        label: "Go to inbox",
        shortcut: [["g"], ["i"]],
        goto: "inbox",
      },
      {
        icon: "dot",
        label: "Go to focus",
        shortcut: [["g"], ["f"]],
        goto: "focus",
      },
    ],
  },
];

const defaultSubmenus = {
  inbox: {
    title: "Inbox",
    items: [
      {
        label: "This is not spam!",
        onSelect: () => jest.fn(),
      },
      {
        label: "Remember to hire Karol.",
        onSelect: () => jest.fn(),
      },
      {
        label: "Hey, how is it going?",
        onSelect: () => jest.fn(),
      },
    ],
    groups: [
      {
        label: "Mail actions...",
        items: [
          {
            icon: "pencil",
            label: "Create new message",
            shortcut: [["m"]],
            onSelect: () => jest.fn(),
          },
          {
            icon: "reschedule",
            label: "Check for new mail",
            shortcut: [["r"]],
            onSelect: () => jest.fn(),
          },
        ],
      },
      {
        label: "Go to...",
        items: [
          {
            icon: "dot",
            label: "Go to focus",
            shortcut: [["g"], ["f"]],
            goto: "../focus",
          },
          {
            label: "Go to main menu",
            shortcut: [["g"], ["m"]],
            goto: "..",
          },
        ],
      },
    ],
  },
  focus: {
    title: "Focus",
    items: [
      {
        icon: "bolt",
        label: "Focus on productivity",
        shortcut: [["f"], ["p"]],
        onSelect: () => jest.fn(),
      },
      {
        icon: "bolt",
        label: "Focus on work life balance",
        shortcut: [["f"], ["w", "l"], ["b"]],
        onSelect: () => jest.fn(),
      },
      {
        icon: "bolt",
        label: "Focus on quality",
        shortcut: [["f"], ["q"]],
        onSelect: () => jest.fn(),
      },
    ],
    groups: [
      {
        label: "Go to...",
        items: [
          {
            icon: "inbox",
            label: "Go to inbox",
            shortcut: [["g"], ["f"]],
            goto: "../inbox",
          },
          {
            label: "Go to main menu",
            shortcut: [["g"], ["m"]],
            goto: "..",
          },
        ],
      },
    ],
  },
};

const renderMenu = (props: Partial<MenuProps>) =>
  render(
    <Menu
      title={props.title || defaultTitle}
      items={props.items || defaultItems}
      groups={props.groups || defaultGroups}
      submenus={props.submenus || defaultSubmenus}
    />
  );

afterEach(() => {
  cleanup();
});

test("renders correctly", () => {
  const { getByTestId } = renderMenu({});
  const menuDiv = getByTestId("Laserfocus-Menu");
  expect(menuDiv).toMatchSnapshot();
});

test("gives focus to search box on initial render", () => {
  const { getByTestId } = renderMenu({});
  const menuDiv = getByTestId("Laserfocus-Menu");
  const input = menuDiv.querySelector("input") as HTMLInputElement;
  expect(window.document.activeElement).toBe(input);
});

test("removes focus from search box on key up/down", () => {
  const { getByTestId } = renderMenu({});
  const menuDiv = getByTestId("Laserfocus-Menu");
  const input = menuDiv.querySelector("input") as HTMLInputElement;
  fireEvent.keyUp(window, { key: "ArrowUp" });
  expect(window.document.activeElement).not.toBe(input);
  fireEvent.focus(input);
  fireEvent.keyUp(window, { key: "ArrowDown" });
  expect(window.document.activeElement).not.toBe(input);
});

test("removes focus from search box on mousemove on items", () => {
  const { getByTestId } = renderMenu({});
  const menuDiv = getByTestId("Laserfocus-Menu");
  const input = menuDiv.querySelector("input") as HTMLInputElement;
  const item = menuDiv.querySelector(".Laserfocus-Menu-Item") as HTMLDivElement;
  fireEvent.mouseMove(item);
  expect(window.document.activeElement).not.toBe(input);
});

test("restores focus to search box on left arrow", () => {
  const { getByTestId } = renderMenu({});
  const menuDiv = getByTestId("Laserfocus-Menu");
  const input = menuDiv.querySelector("input") as HTMLInputElement;
  fireEvent.keyUp(window, { key: "ArrowDown" });
  fireEvent.keyUp(window, { key: "ArrowLeft" });
  expect(window.document.activeElement).toBe(input);
});

test("filters when user types into search box", () => {
  const { getByTestId, getByText } = renderMenu({});
  const menuDiv = getByTestId("Laserfocus-Menu");
  const input = menuDiv.querySelector("input") as HTMLInputElement;
  fireEvent.change(input, { target: { value: "log" } });
  expect(menuDiv).toMatchSnapshot();
  expect(getByText("Search results for: log")).toBeInTheDocument();
  expect(menuDiv.querySelectorAll(".Laserfocus-Menu-Item").length).toEqual(1);
});

test("follows goto in item correctly", () => {
  const { getByTestId, getByText } = renderMenu({});
  const menuDiv = getByTestId("Laserfocus-Menu");
  fireEvent.keyDown(window, { key: "g" });
  fireEvent.keyUp(window, { key: "g" });
  fireEvent.keyDown(window, { key: "i" });
  fireEvent.keyUp(window, { key: "i" });
  expect(menuDiv).toMatchSnapshot();
  expect(getByText("Inbox")).toBeInTheDocument();
  expect(menuDiv.querySelectorAll(".Laserfocus-Menu-Item").length).toEqual(7);
});
