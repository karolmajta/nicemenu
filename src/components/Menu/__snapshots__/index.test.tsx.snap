// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filters when user types into search box 1`] = `
<div
  class="Laserfocus-Menu sc-eCApGN kJvQwO"
  data-testid="Laserfocus-Menu"
>
  <div
    class="menu-title"
  >
    <div
      class="menu-title-text"
    >
      Search results for: log
    </div>
  </div>
  <div
    class="Laserfocus-Menu-SearchBox sc-hKFyIo ivSqkU"
    data-testid="Laserfocus-Menu-SearchBox"
  >
    <input
      placeholder="Search..."
      type="text"
      value="log"
    />
  </div>
  <div
    class="Laserfocus-Menu-Item sc-gtssRu iLlUnN"
    data-testid="Laserfocus-Menu-Item"
  >
    <div
      class="Laserfocus-Menu-Item-Icon sc-bdnylx uunbO"
      data-testid="Laserfocus-Menu-Item-Icon"
    >
      <img
        alt="bolt"
        src="bolt.svg"
      />
    </div>
    <div
      class="item-label"
    >
      Log an activity
    </div>
    <div
      class="item-shortcut"
    >
      <div
        class="item-shortcut-parts"
      >
        <div
          class="item-shortcut-step"
        >
          c
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`follows goto in item correctly 1`] = `
<div
  class="Laserfocus-Menu sc-eCApGN kJvQwO"
  data-testid="Laserfocus-Menu"
>
  <div
    class="menu-title"
  >
    <div
      class="menu-title-text"
    >
      Inbox
    </div>
  </div>
  <div
    class="Laserfocus-Menu-SearchBox sc-hKFyIo ivSqkU"
    data-testid="Laserfocus-Menu-SearchBox"
  >
    <input
      placeholder="Press arrow left to search or type command..."
      type="text"
      value=""
    />
  </div>
  <div
    class="Laserfocus-Menu-Item highlighted sc-gtssRu iLlUnN"
    data-testid="Laserfocus-Menu-Item"
  >
    <div
      class="icon-placeholder"
    />
    <div
      class="item-label"
    >
      This is not spam!
    </div>
  </div>
  <div
    class="Laserfocus-Menu-Item sc-gtssRu iLlUnN"
    data-testid="Laserfocus-Menu-Item"
  >
    <div
      class="icon-placeholder"
    />
    <div
      class="item-label"
    >
      Remember to hire Karol.
    </div>
  </div>
  <div
    class="Laserfocus-Menu-Item sc-gtssRu iLlUnN"
    data-testid="Laserfocus-Menu-Item"
  >
    <div
      class="icon-placeholder"
    />
    <div
      class="item-label"
    >
      Hey, how is it going?
    </div>
  </div>
  <div
    class="Laserfocus-Menu-Group sc-dlnjPT fMMgoZ"
    data-testid="Laserfocus-Menu-Group"
  >
    <div
      class="group-label"
    >
      <div
        class="group-label-text"
      >
        Mail actions...
      </div>
    </div>
    <div
      class="Laserfocus-Menu-Item sc-gtssRu iLlUnN"
      data-testid="Laserfocus-Menu-Item"
    >
      <div
        class="Laserfocus-Menu-Item-Icon sc-bdnylx uunbO"
        data-testid="Laserfocus-Menu-Item-Icon"
      >
        <img
          alt="pencil"
          src="pencil.svg"
        />
      </div>
      <div
        class="item-label"
      >
        Create new message
      </div>
      <div
        class="item-shortcut"
      >
        <div
          class="item-shortcut-parts"
        >
          <div
            class="item-shortcut-step"
          >
            m
          </div>
        </div>
      </div>
    </div>
    <div
      class="Laserfocus-Menu-Item sc-gtssRu iLlUnN"
      data-testid="Laserfocus-Menu-Item"
    >
      <div
        class="Laserfocus-Menu-Item-Icon sc-bdnylx uunbO"
        data-testid="Laserfocus-Menu-Item-Icon"
      >
        <img
          alt="reschedule"
          src="reschedule.svg"
        />
      </div>
      <div
        class="item-label"
      >
        Check for new mail
      </div>
      <div
        class="item-shortcut"
      >
        <div
          class="item-shortcut-parts"
        >
          <div
            class="item-shortcut-step"
          >
            r
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="Laserfocus-Menu-Group sc-dlnjPT fMMgoZ"
    data-testid="Laserfocus-Menu-Group"
  >
    <div
      class="group-label"
    >
      <div
        class="group-label-text"
      >
        Go to...
      </div>
    </div>
    <div
      class="Laserfocus-Menu-Item sc-gtssRu iLlUnN"
      data-testid="Laserfocus-Menu-Item"
    >
      <div
        class="Laserfocus-Menu-Item-Icon sc-bdnylx uunbO"
        data-testid="Laserfocus-Menu-Item-Icon"
      >
        <img
          alt="dot"
          src="dot.svg"
        />
      </div>
      <div
        class="item-label"
      >
        Go to focus
      </div>
      <div
        class="item-shortcut"
      >
        <div
          class="item-shortcut-parts"
        >
          <div
            class="item-shortcut-step"
          >
            g
          </div>
          <div
            class="item-shortcut-part-separator"
          >
            then
          </div>
        </div>
        <div
          class="item-shortcut-parts"
        >
          <div
            class="item-shortcut-step"
          >
            f
          </div>
        </div>
      </div>
    </div>
    <div
      class="Laserfocus-Menu-Item sc-gtssRu iLlUnN"
      data-testid="Laserfocus-Menu-Item"
    >
      <div
        class="icon-placeholder"
      />
      <div
        class="item-label"
      >
        Go to main menu
      </div>
      <div
        class="item-shortcut"
      >
        <div
          class="item-shortcut-parts"
        >
          <div
            class="item-shortcut-step"
          >
            g
          </div>
          <div
            class="item-shortcut-part-separator"
          >
            then
          </div>
        </div>
        <div
          class="item-shortcut-parts"
        >
          <div
            class="item-shortcut-step"
          >
            m
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders correctly 1`] = `
<div
  class="Laserfocus-Menu sc-eCApGN kJvQwO"
  data-testid="Laserfocus-Menu"
>
  <div
    class="menu-title"
  >
    <div
      class="menu-title-text"
    >
      Main menu
    </div>
  </div>
  <div
    class="Laserfocus-Menu-SearchBox sc-hKFyIo ivSqkU"
    data-testid="Laserfocus-Menu-SearchBox"
  >
    <input
      placeholder="Search..."
      type="text"
      value=""
    />
  </div>
  <div
    class="Laserfocus-Menu-Item sc-gtssRu iLlUnN"
    data-testid="Laserfocus-Menu-Item"
  >
    <div
      class="Laserfocus-Menu-Item-Icon sc-bdnylx uunbO"
      data-testid="Laserfocus-Menu-Item-Icon"
    >
      <img
        alt="bolt"
        src="bolt.svg"
      />
    </div>
    <div
      class="item-label"
    >
      Log an activity
    </div>
    <div
      class="item-shortcut"
    >
      <div
        class="item-shortcut-parts"
      >
        <div
          class="item-shortcut-step"
        >
          c
        </div>
      </div>
    </div>
  </div>
  <div
    class="Laserfocus-Menu-Item sc-gtssRu iLlUnN"
    data-testid="Laserfocus-Menu-Item"
  >
    <div
      class="Laserfocus-Menu-Item-Icon sc-bdnylx uunbO"
      data-testid="Laserfocus-Menu-Item-Icon"
    >
      <img
        alt="reschedule"
        src="reschedule.svg"
      />
    </div>
    <div
      class="item-label"
    >
      Set a reminder
    </div>
    <div
      class="item-shortcut"
    >
      <div
        class="item-shortcut-parts"
      >
        <div
          class="item-shortcut-step"
        >
          h
        </div>
      </div>
    </div>
  </div>
  <div
    class="Laserfocus-Menu-Item sc-gtssRu iLlUnN"
    data-testid="Laserfocus-Menu-Item"
  >
    <div
      class="Laserfocus-Menu-Item-Icon sc-bdnylx uunbO"
      data-testid="Laserfocus-Menu-Item-Icon"
    >
      <img
        alt="loader"
        src="loader.svg"
      />
    </div>
    <div
      class="item-label"
    >
      Change status
    </div>
    <div
      class="item-shortcut"
    >
      <div
        class="item-shortcut-parts"
      >
        <div
          class="item-shortcut-step"
        >
          Shift
        </div>
        <div
          class="item-shortcut-step"
        >
          s
        </div>
      </div>
    </div>
  </div>
  <div
    class="Laserfocus-Menu-Item sc-gtssRu iLlUnN"
    data-testid="Laserfocus-Menu-Item"
  >
    <div
      class="Laserfocus-Menu-Item-Icon sc-bdnylx uunbO"
      data-testid="Laserfocus-Menu-Item-Icon"
    >
      <img
        alt="pencil"
        src="pencil.svg"
      />
    </div>
    <div
      class="item-label"
    >
      Edit contact
    </div>
    <div
      class="item-shortcut"
    >
      <div
        class="item-shortcut-parts"
      >
        <div
          class="item-shortcut-step"
        >
          e
        </div>
      </div>
    </div>
  </div>
  <div
    class="Laserfocus-Menu-Group sc-dlnjPT fMMgoZ"
    data-testid="Laserfocus-Menu-Group"
  >
    <div
      class="group-label"
    >
      <div
        class="group-label-text"
      >
        Go to...
      </div>
    </div>
    <div
      class="Laserfocus-Menu-Item sc-gtssRu iLlUnN"
      data-testid="Laserfocus-Menu-Item"
    >
      <div
        class="Laserfocus-Menu-Item-Icon sc-bdnylx uunbO"
        data-testid="Laserfocus-Menu-Item-Icon"
      >
        <img
          alt="inbox"
          src="inbox.svg"
        />
      </div>
      <div
        class="item-label"
      >
        Go to inbox
      </div>
      <div
        class="item-shortcut"
      >
        <div
          class="item-shortcut-parts"
        >
          <div
            class="item-shortcut-step"
          >
            g
          </div>
          <div
            class="item-shortcut-part-separator"
          >
            then
          </div>
        </div>
        <div
          class="item-shortcut-parts"
        >
          <div
            class="item-shortcut-step"
          >
            i
          </div>
        </div>
      </div>
    </div>
    <div
      class="Laserfocus-Menu-Item sc-gtssRu iLlUnN"
      data-testid="Laserfocus-Menu-Item"
    >
      <div
        class="Laserfocus-Menu-Item-Icon sc-bdnylx uunbO"
        data-testid="Laserfocus-Menu-Item-Icon"
      >
        <img
          alt="dot"
          src="dot.svg"
        />
      </div>
      <div
        class="item-label"
      >
        Go to focus
      </div>
      <div
        class="item-shortcut"
      >
        <div
          class="item-shortcut-parts"
        >
          <div
            class="item-shortcut-step"
          >
            g
          </div>
          <div
            class="item-shortcut-part-separator"
          >
            then
          </div>
        </div>
        <div
          class="item-shortcut-parts"
        >
          <div
            class="item-shortcut-step"
          >
            f
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
